src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp

```diff
@@ -15,751 +15,878 @@
+                    me->RemoveAurasDueToSpell(SPELL_RANDOM_AGGRO_PERIODIC);
+                    me->AttackStop();
+                    me->SetReactState(REACT_PASSIVE);
+                    me->GetMotionMaster()->MoveLand(POINT_ID_LAND, { me->GetPositionX(), me->GetPositionY(), me->GetFloorZ() }, 3.75f);
+                    break;
+                default:
+                    break;
+            }
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+        }
 
-            private:
-                ObjectGuid m_TriggerGUID;
-        };
+        DoMeleeAttackIfReady();
+    }
 
-        CreatureAI* GetAI(Creature* creature) const override
-        {
-            return GetUlduarAI<boss_stormcaller_brundirAI>(creature);
-        }
+private:
+    EventMap _events;
+    InstanceScript* _instance;
+    SummonList _summons;
+    std::vector<Position> _potentialWaypoints;
 };
 
-class spell_shield_of_runes : public SpellScriptLoader
+struct npc_assembly_lightning_elemental : public ScriptedAI
 {
-    public:
-        spell_shield_of_runes() : SpellScriptLoader("spell_shield_of_runes") { }
-
-        class spell_shield_of_runes_AuraScript : public AuraScript
-        {
-            PrepareAuraScript(spell_shield_of_runes_AuraScript);
-
-            void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
-            {
-                if (Unit* caster = GetCaster())
-                    if (GetTargetApplication()->GetRemoveMode() != AURA_REMOVE_BY_EXPIRE)
-                        caster->CastSpell(caster, SPELL_SHIELD_OF_RUNES_BUFF, false);
-            }
-
-            void Register() override
-            {
-                 AfterEffectRemove += AuraEffectRemoveFn(spell_shield_of_runes_AuraScript::OnRemove, EFFECT_0, SPELL_AURA_SCHOOL_ABSORB, AURA_EFFECT_HANDLE_REAL);
-            }
-        };
+    npc_assembly_lightning_elemental(Creature* creature) : ScriptedAI(creature) { }
+
+    void InitializeAI() override
+    {
+        if (InstanceScript* instance = me->GetInstanceScript())
+            if (Creature* molgeim = instance->GetCreature(DATA_RUNEMASTER_MOLGEIM))
+                if (molgeim->IsAIEnabled())
+                    molgeim->AI()->JustSummoned(me);
+    }
+
+    void JustAppeared() override
+    {
+        if (Unit* target = me->SelectNearestTarget(100.f, true))
+            me->EngageWithTarget(target);
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        me->DespawnOrUnsummon(2s);
+    }
+};
 
-        AuraScript* GetAuraScript() const override
-        {
-            return new spell_shield_of_runes_AuraScript();
-        }
+class spell_shield_of_runes : public AuraScript
+{
+    PrepareAuraScript(spell_shield_of_runes);
+
+    void AfterRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (Unit* caster = GetCaster())
+            if (GetTargetApplication()->GetRemoveMode() != AURA_REMOVE_BY_EXPIRE)
```

Comment:
https://wowwiki.fandom.com/wiki/Assembly_of_Iron
>Shield of Runes - Shield absorbs up to 50,000 heroic 20,000 normal damage. If the shield absorbs the total amount this gives Moldeim a 50% damage increase for 15 seconds. Should be purged, dispelled or spellstolen.

in case of remove by dispel shouldn't get the shield of runes buff

Reply:
this can fix it, not sure if can be the proper way:
```diff
--- a/src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp
+++ b/src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp
@@
 #include "ObjectAccessor.h"
 #include "ScriptedCreature.h"
 #include "SpellAuras.h"
+#include "SpellAuraEffects.h"
 #include "SpellMgr.h"
 #include "SpellScript.h"
 #include "Spell.h"
@@ class spell_shield_of_runes : public AuraScript
 {
     PrepareAuraScript(spell_shield_of_runes);
 
-    void AfterRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    void AfterRemove(AuraEffect const* aurEff, AuraEffectHandleModes /*mode*/)
     {
         if (Unit* caster = GetCaster())
-            if (GetTargetApplication()->GetRemoveMode() != AURA_REMOVE_BY_EXPIRE)
+            if (GetTargetApplication()->GetRemoveMode() != AURA_REMOVE_BY_EXPIRE && aurEff->GetAmount() <= 0)
                 caster->CastSpell(caster, SPELL_SHIELD_OF_RUNES_BUFF, false);
     }
 ```